src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -97,56 +118,130 @@ internal class PayloadSenderV2 : IPayloadSender, IDisposable
+			}
 		}
 
 		private CancellationTokenSource _batchBlockReceiveAsyncCts;
+		private long _eventQueueCount;
 
-		public void QueueTransaction(ITransaction transaction)
-		{
-			var res = _eventQueue.Post(transaction);
-			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Transaction to the queue, {Transaction}"
-					: "Transaction added to the queue, {Transaction}", transaction);
+		public void QueueTransaction(ITransaction transaction) => EnqueueEvent(transaction, "Transaction");
 
-			_eventQueue.TriggerBatch();
-		}
+		public void QueueSpan(ISpan span) => EnqueueEvent(span, "Span");
 
-		public void QueueSpan(ISpan span)
-		{
-			var res = _eventQueue.Post(span);
-			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Span to the queue, {Span}"
-					: "Span added to the queue, {Span}", span);
-		}
+		public void QueueMetrics(IMetricSet metricSet) => EnqueueEvent(metricSet, "MetricSet");
 
-		public void QueueMetrics(IMetricSet metricSet)
-		{
-			var res = _eventQueue.Post(metricSet);
-			_logger.Debug()
-				?.Log(!res
-					? "Failed adding MetricSet to the queue, {MetricSet}"
-					: "MetricSet added to the queue, {MetricSet}", metricSet);
-		}
+		public void QueueError(IError error) => EnqueueEvent(error, "Error");
 
-		public void QueueError(IError error)
+		internal bool EnqueueEvent(object eventObj, string dbgEventKind)
 		{
-			var res = _eventQueue.Post(error);
+			// Enforce _maxQueueEventCount manually instead of using BatchBlock's BoundedCapacity
+			// because of the issue of Post returning false when TriggerBatch is in progress. For more details see
+			// https://stackoverflow.com/questions/35626955/unexpected-behaviour-tpl-dataflow-batchblock-rejects-items-while-triggerbatch
+			var newEventQueueCount = Interlocked.Increment(ref _eventQueueCount);
+			if (newEventQueueCount > _maxQueueEventCount)
+			{
+				_logger.Debug()
+					?.Log("Queue reached max capacity - " + dbgEventKind + " will be discarded. "
+						+ " newEventQueueCount: {EventQueueCount}."
+						+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+						, newEventQueueCount, eventObj);
+				Interlocked.Decrement(ref _eventQueueCount);
+				return false;
+			}
+
+			var enqueuedSuccessfully = _eventQueue.Post(eventObj);
+			if (!enqueuedSuccessfully)
+			{
+				_logger.Debug()
+					?.Log("Failed to enqueue " + dbgEventKind + "."
+						+ " newEventQueueCount: {EventQueueCount}."
+						+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+						, newEventQueueCount, eventObj);
+				Interlocked.Decrement(ref _eventQueueCount);
+				return false;
+			}
+
 			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Error to the queue, {Error}"
-					: "Error added to the queue, {Error}", error);
+				?.Log("Enqueued " + dbgEventKind + "."
+					+ " newEventQueueCount: {EventQueueCount}."
+					+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+					, newEventQueueCount, eventObj);
+
+			if (_flushInterval == TimeSpan.Zero) _eventQueue.TriggerBatch();
```

Comment:
`_eventQueue.TriggerBatch()` will be called for each event only when `FlushInterval` is `0` which is exactly as documented:
> If `FlushInterval` is set to `0` (or `0s`, `0ms`, etc.) and there's no event sending operation still in progress then the Agent won't hold events in the queue and instead will send them immediately.

But by default `FlushInterval` is not `0`, it's `10s` so that line won't be called. Do I miss something?

Reply:
Ok, fine by me. Given that in the default case this won't be executed I think it's fine.